src/solver.rs

```diff
@@ -93,50 +92,54 @@ pub fn resolve<P: Package, V: Version>(
             .should_cancel()
             .map_err(|err| PubGrubError::ErrorShouldCancel(err))?;
 
-        state.unit_propagation(next)?;
+        state.unit_propagation(next.clone())?;
 
-        // Pick the next package.
-        let (p, term) = match state.partial_solution.pick_package(dependency_provider)? {
-            None => {
-                return state.partial_solution.extract_solution().ok_or_else(|| {
-                    PubGrubError::Failure(
-                        "How did we end up with no package to choose but no solution?".into(),
-                    )
-                })
-            }
-            Some(x) => x,
-        };
-        next = p.clone();
-        let available_versions =
-            dependency_provider
-                .list_available_versions(&p)
-                .map_err(|err| PubGrubError::ErrorRetrievingVersions {
-                    package: p.clone(),
-                    source: err,
-                })?;
+        let iter = state.partial_solution.potential_packages();
+        if iter.is_none() {
+            drop(iter);
```

Comment:
The borrow checker did not like using a match on iter. This `if ... is_none ... drop` is a workaround. I believe this is a case where [`Polonius`](https://github.com/rust-lang/polonius) could help, when and if it lands in rustc.